### ISSUE TITLE
Empty raw block at the end of a page

### ISSUE THREAD

**Issue by Narukara** [typst/typst#2914]
### Description

I want to put some code in a block with background color. But I found that sometimes, at the end of a page, there would be a block with color but no content. It looks weird. I'm not sure if this is a bug, but blocks containing normal text don't seem to cause this problem.

A example:

````typst
#show raw.where(block: true): it => [
  #block(fill: luma(230), inset: 10pt, width: 100%)[
    #it
  ]
]

#lorem(630)

```rust
fn main() {
    println!("Hello, world!");
}
```
````
![图片](https://github.com/typst/typst/assets/53737377/d7847d62-6969-4fdf-be45-c7e64a48bb37)

typst version: 0.10.0 (70ca0d25)

### Operating system

Linux

### Typst version

- [X] I am using the latest version of Typst

**Comment by LaurenzV** [comment]
Seems like for some reason it tries to split the raw block up by breaking it, even though the first line doesn't fit anymore. I guess in this particular case you can workaround it by adding `breakable: false` to the block.

**Comment by Narukara** [comment]
> Seems like for some reason it tries to split the raw block up by breaking it, even though the first line doesn't fit anymore. I guess in this particular case you can workaround it by adding `breakable: false` to the block.

Adding `breakable: false` does solve the problem. But if I happen to have very long code, this may cause more problems. :cry: 

like this:

![图片](https://github.com/typst/typst/assets/53737377/f5a287dc-fe97-4faf-9845-568c21bdf467)

**Comment by laurmaedje** [comment]
Rawless reproduction:
```typ
#lorem(630)
#block(
  fill: luma(230), 
  inset: 10pt,
  width: 100%, 
  block[a \ b],
)
```
(Raw also produces an internal block.)

I think the cause is that the first frame of the outer block doesn't end up empty because of the inner block, which causes the ad-hoc solution of preventing background to fail.

**Comment by Andrew15-5** [comment]
I used `showybox` for an easier outer block customization. But when there is a title and this bug appears it looks even worse:

![image](https://github.com/typst/typst/assets/37143421/bc01f5d5-ae29-4a98-98df-3af8b8a55523)

This is why I have to manually unbreak some blocks. The annoying part is that not only I have a total of **more** than 200 of these, but also the need for unbreak depends on the page format/size.

```typ
#clips-function-example("open", breakable: ("a4": false)
```

**Comment by jbirnick** [comment]
Per user request I will post a workaround here:
You can edit the block properties directly with a `set` rule (inside the corresponding `show` rule) instead of wrapping it in another block.

So for the original example:
```typ
#show raw.where(block: true): set block(fill: luma(230), inset: 10pt, width: 100%)

#lorem(630)

```rust
fn main() {
    println!("Hello, world!");
}
```

**Comment by Andrew15-5** [comment]
> But I don't know whether [@Andrew15-5](https://github.com/Andrew15-5) 's problem is solved. And I am not sure if I should close this issue.

I don't think it is. You shouldn't, as workarounds aren't a proper fix. Otherwise this would've been closed a long time ago.

**Comment by Andrew15-5** [comment]
@bluss found something interesting at https://forum.typst.app/t/math-theorem-package-with-smart-break-between-pages/4003.

First it's [this](https://forum.typst.app/t/math-theorem-package-with-smart-break-between-pages/4003/4?u=andrew):

```typ
#show raw.where(block: true): it => [
  #block(fill: luma(230), inset: 10pt, width: 100%)[
    #set block(sticky: true)
    #it
  ]
]
```

That works like an unbreakable block, that will still break if the content is longer than one page. But it will not flow correctly. So ultimately it's not a fix, though I have questions:
- How come one `RawElem`'s `block.sticky` makes it that all `RawLine`s are glued together, even if they are just `TextElem`?
- This could be useful if you want to have a "soft unbreakable" block, that will only break, if doesn't fit on one page (with minimal amount of breaking points). Maybe a new built-in feature or a package or note in documentation?

The second, more promising workaround is [this](https://forum.typst.app/t/math-theorem-package-with-smart-break-between-pages/4003/15?u=andrew):

```typ
#show block.where(sticky: false): it => {
  block(sticky: true, height: 0pt, below: 0pt)[]
  it
}
```

Which can be used like [this](https://forum.typst.app/t/math-theorem-package-with-smart-break-between-pages/4003/17?u=andrew):

```typ
#let use-workaround = false
#let use-workaround = true

#let code(body) = block(stroke: 1pt, width: 100%, inset: 5pt, {
  if use-workaround { block(sticky: true, spacing: 0pt) }
  body
})
```

From what I tested, it does work like I would expect, didn't find any issues with it. I think it exploits some logic in Typst to actually make the layout _better_.

**Comment by knuesel** [comment]
Instead of skipping fill/stroke for empty parts, isn't the proper fix to move the block start to the next page? Otherwise one can get surprising behavior like this:

```typ
#set page(width: 8cm, height: 10cm)
#let b = block(fill: blue, width: 100%, height: 6cm, inset: 1cm)[#lorem(8)]

#b
#b
```

![Image](https://github.com/user-attachments/assets/c571b122-6c23-4b63-b6a4-963ffde9f71d)

(With `block(fill: blue, height: 6cm)` I would expect to see  those 6cm of blue but the second block looks shorter.)
